viewing_party/party.py

```diff
@@ -1,23 +1,151 @@
+        rating_sum += movie["rating"]
+    
+    length = len(user_data["watched"])
+    return rating_sum / length
+
+# 2nd function in wave_02
+def get_most_watched_genre(user_data): 
+    
+    if not user_data["watched"]:
+        return None
+    
+    genres = {}
+    for movie in user_data["watched"]: 
+        if movie["genre"] not in genres: 
+            genres[movie["genre"]] = 0
+        else: 
+            genres[movie["genre"]] += 1 
+    
+
+    return max(genres, key = genres.get)
+
 
 
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
+#1st function in wave_03
+def get_unique_watched(user_data):
+    user_unique_movies = []
+    my_friends_movies = set() 
 
+    for movie in user_data["friends"]:
+        for num in range(len(movie["watched"])): 
+            my_friends_movies.add(movie["watched"][num]["title"])
+    
+    for pelicula in user_data["watched"]:
+        
+        if pelicula["title"] not in my_friends_movies: 
+            user_unique_movies.append(pelicula)
         
+    return user_unique_movies
+
+# 2nd function in wave_03 
+def get_friends_unique_watched(user_data):
+    
+    friends_unique_movies = {}
+    user_movies = {}
+
+
+    for movie in user_data["watched"]:
+        user_movies[movie["title"]] = movie
+
+    for pelicula in user_data["friends"]:
+        for num in range(len(pelicula["watched"])):
+            title = pelicula["watched"][num]["title"]
+            if title not in user_movies and title not in friends_unique_movies: 
+                friends_unique_movies[title] = pelicula["watched"][num]
+    
+    result = list(friends_unique_movies.values())
+
+    return result
+            
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
+# 1st function for wave  04 
+def get_available_recs(user_data): 
+    movie_recs = []
+    unique_movies = get_friends_unique_watched(user_data)
+
+    for movie in unique_movies: 
+        if movie["host"] in user_data["subscriptions"]: 
+            movie_recs.append(movie)
+    
+    return movie_recs
+
 
 # -----------------------------------------
 # ------------- WAVE 5 --------------------
 # -----------------------------------------
+# 1st function in wave 05 
+def get_new_rec_by_genre(user_data):
+    
+    movie_recs = []
+    genre = get_most_watched_genre(user_data)
+    friends_unique_movies = get_friends_unique_watched(user_data)
+
+    for movie in friends_unique_movies: 
+        if movie["genre"] == genre:     
+            movie_recs.append(movie)
+    
+    return movie_recs
+
+# 2nd function in wave 05
+def get_rec_from_favorites(user_data): 
+    user_unique_movies = get_unique_watched(user_data)
+    movie_recs = []
 
+    for movie in user_data["favorites"]: 
+        if movie in user_unique_movies: 
+            movie_recs.append(movie)

```

Comment:
[perf] What's the big-O complexity of this loop? (i.e. what is the type of `user_unique_movies`, and what is the consequence of the `in` operator?)